Add `embedded-hal-async` support

hawkw opened the pull request:
This commit adds support for the `embedded-hal-async` crate in addition
to `embedded-hal`. I've done this by adding a separate `AsyncSht4x`
type, based on the assumption that most projects won't need to use both
the blocking `embedded-hal` traits and the `embedded-hal-async` traits
at the same time, and providing `async fn` methods on a separate type
with the same names as the blocking ones seemed a bit nicer than having
one type that has both `fn measure` and `async fn measure_async` and so
on.

Support for `embedded-hal-async` is gated behind the
`embedded-hal-async` feature flag, so the dependency is not enabled by
default.

Note that this branch depends on my PR https://github.com/sirhcel/sht4x/pull/6, which updates this crate to
use `embedded-hal` v1.0, and currently contains the commit from that
change as well. Once https://github.com/sirhcel/sht4x/pull/6 has merged, this branch will need to be rebased
onto the main branch.

It also depends on my upstream PR adding `embedded-hal-async` support to
`sensirion-i2c-rs`, https://github.com/Sensirion/sensirion-i2c-rs/pull/30, which has been
[merged], but hasn't been published to crates.io yet. Currently, this
branch adds a Cargo `[patch]` to use a Git dep on `sensirion-i2c-rs`.
So, this change cannot be released to crates.io until upstream publishes
a new release of `sensirion-i2c-rs`. Hopefully they do that soon! :)

[merged]: https://github.com/Sensirion/sensirion-i2c-rs/commit/f7b9f3a81b777bc6e6b2f0acb4c1ef9c57dfa06d

hawkw committed:
Add `embedded-hal-async` support

This commit adds support for the `embedded-hal-async` crate in addition
to `embedded-hal`. I've done this by adding a separate `AsyncSht4x`
type, based on the assumption that most projects won't need to use both
the blocking `embedded-hal` traits and the `embedded-hal-async` traits
at the same time, and providing `async fn` methods on a separate type
with the same names as the blocking ones seemed a bit nicer than having
one type that has both `fn measure` and `async fn measure_async` and so
on.

Support for `embedded-hal-async` is gated behind the
`embedded-hal-async` feature flag, so the dependency is not enabled by
default.

Note that this branch depends on my PR #6, which updates this crate to
use `embedded-hal` v1.0, and currently contains the commit from that
change as well. Once #6 has merged, this branch will need to be rebased
onto the main branch.

It also depends on my upstream PR adding `embedded-hal-async` support to
`sensirion-i2c-rs`, sensirion/sensirion-i2c-rs#30, which has been
[merged], but hasn't been published to crates.io yet. Currently, this
branch adds a Cargo `[patch]` to use a Git dep on `sensirion-i2c-rs`.
So, this change cannot be released to crates.io until upstream publishes
a new release of `sensirion-i2c-rs`. Hopefully they do that soon! :)

[merged]: https://github.com/Sensirion/sensirion-i2c-rs/commit/f7b9f3a81b777bc6e6b2f0acb4c1ef9c57dfa06d

hawkw committed:
Update to `sensirion-i2c` v0.4

Now that my changes have been published upstream, we can remove the Git
patch. This PR should now be okay to merge!

hawkw committed:
Rename async driver to `Sht4xAsync`

This way, the naming scheme is more consistent with the async driver
I added to the `sgp30` crate. See:
dbrgn/sgp30-rs@cb769b04bb07b8daa5ab0a807ad01849804534bf

bbustin commented:
Please also see my pull request which added support for embedded-hal-1.0 and async support gated behind a feature: https://github.com/sirhcel/sht4x/pull/5

sirhcel committed:
Streamline error type parameter name

It's about the error and not the I2C implementation type.

sirhcel committed:
Bump depenencies to latest release

sirhcel committed:
Streamline async driver module name too

With the driver struct already named Sht4xAsync, its source file should
follow the same order.

sirhcel committed:
Share extracting serial number from response

Sensor data extraction is already shared and the serial number
extraction code should be shared as well.

sirhcel committed:
Make extracting payload from response more explicit

It's written in the sensors data sheet but I did not remember this after
looking into the code after quite some time.

sirhcel committed:
Factor out response data processing to responses.rs

sirhcel commented:
Thank you very much for your PR @hawkw! I took me a while to wrap my head around how to provide a sync and async driver variant from a single crate. I wished that more code could be shared but this seems not conveniently possible as of now.

hawkw commented:
> I wished that more code could be shared but this seems not conveniently possible as of now.

Yeah, I also wished that, but unfortunately the story for this sort of thing is still a bit rough. Thanks for taking the time to merge my PR!
